flocker/control/_persistence.py

```diff
@@ -306,6 +310,77 @@ def _cached_dfs_serialize(input_object):
 
     return result
 
+# A couple tokens that are used below in the generation hash.
+_NULLSET_TOKEN = mmh3_hash_bytes(b'NULLSET')
+_MAPPING_TOKEN = mmh3_hash_bytes(b'MAPPING')
+_STR_TOKEN = mmh3_hash_bytes(b'STRING')
+
+_generation_hash_cache = WeakKeyDictionary()
+
+
+def generation_hash(input_object):
+    """
+    This computes the mmh3 hash for an input object, providing a consistent
+    hash of deeply persistent objects across python nodes and implementations.
+
+    :returns: An mmh3 hash of input_object.
+    """
+    # Ensure this is a quick function for basic types:
+    # Note that ``type(x) in frozenset([str, int])`` is faster than
+    # ``isinstance(x, (str, int))``.
+    input_type = type(input_object)
+    if (
+            input_object is None or
+            input_type in _BASIC_JSON_TYPES
+    ):
+        if input_type == unicode or input_type == bytes:
+            # Add a token to identify this as a string.
+            object_to_process = b''.join([_STR_TOKEN, bytes(input_object)])
+        else:
+            # For non-string objects, just hash the JSON encoding.
+            object_to_process = dumps(input_object)
+        return mmh3_hash_bytes(object_to_process)
+
+    is_pyrsistent = _is_pyrsistent(input_object)
+    if is_pyrsistent:
+        cached = _generation_hash_cache.get(input_object, _UNCACHED_SENTINEL)
+        if cached is not _UNCACHED_SENTINEL:
+            return cached
+
+    object_to_process = input_object
+
+    if isinstance(object_to_process, PClass):
+        object_to_process = object_to_process._to_dict()
+
+    if isinstance(object_to_process, Mapping):
+        # Union a mapping token so that empty maps and empty sets have
+        # different hashes.
+        object_to_process = frozenset(object_to_process.iteritems()).union(
+            [_MAPPING_TOKEN]
+        )
+
+    if isinstance(object_to_process, Set):
+        def xor_bytes(a, b):
+            for i in xrange(len(a)):
+                a[i] ^= ord(b[i])
+            return a
+
+        sub_hashes = (generation_hash(x) for x in object_to_process)
+        result = bytes(
+            reduce(xor_bytes, sub_hashes, bytearray(_NULLSET_TOKEN))
+        )
```

Comment:
I wondered if there might be something in the `cryptography` library that does this even faster.